// openclaw/dist/daemon/systemd.js
import { execFile } from "node:child_process";
import fs from "node:fs/promises";
import path from "node:path";
import { promisify } from "node:util";
import { colorize, isRich, theme } from "../terminal/theme.js";
import { formatGatewayServiceDescription, LEGACY_GATEWAY_SYSTEMD_SERVICE_NAMES, resolveGatewaySystemdServiceName, } from "./constants.js";
import { resolveHomeDir } from "./paths.js";
import { parseKeyValueOutput } from "./runtime-parse.js";
import { enableSystemdUserLinger, readSystemdUserLingerStatus, } from "./systemd-linger.js";
import { buildSystemdUnit, parseSystemdEnvAssignment, parseSystemdExecStart, } from "./systemd-unit.js";
const execFileAsync = promisify(execFile);
const toPosixPath = (value) => value.replace(/\\/g, "/");
const formatLine = (label, value) => {
    const rich = isRich();
    return `${colorize(rich, theme.muted, `${label}:`)} ${colorize(rich, theme.command, value)}`;
};
function resolveSystemdUnitPathForName(env, name) {
    const home = toPosixPath(resolveHomeDir(env));
    return path.posix.join(home, ".config", "systemd", "user", `${name}.service`);
}
function resolveSystemdServiceName(env) {
    const override = env.OPENCLAW_SYSTEMD_UNIT?.trim();
    if (override) {
        return override.endsWith(".service") ? override.slice(0, -".service".length) : override;
    }
    return resolveGatewaySystemdServiceName(env.OPENCLAW_PROFILE);
}
function resolveSystemdUnitPath(env) {
    return resolveSystemdUnitPathForName(env, resolveSystemdServiceName(env));
}
export function resolveSystemdUserUnitPath(env) {
    return resolveSystemdUnitPath(env);
}
export { enableSystemdUserLinger, readSystemdUserLingerStatus };
// Unit file parsing/rendering: see systemd-unit.ts
export async function readSystemdServiceExecStart(env) {
    const unitPath = resolveSystemdUnitPath(env);
    try {
        const content = await fs.readFile(unitPath, "utf8");
        let execStart = "";
        let workingDirectory = "";
        const environment = {};
        for (const rawLine of content.split("\n")) {
            const line = rawLine.trim();
            if (!line || line.startsWith("#")) {
                continue;
            }
            if (line.startsWith("ExecStart=")) {
                execStart = line.slice("ExecStart=".length).trim();
            }
            else if (line.startsWith("WorkingDirectory=")) {
                workingDirectory = line.slice("WorkingDirectory=".length).trim();
            }
            else if (line.startsWith("Environment=")) {
                const raw = line.slice("Environment=".length).trim();
                const parsed = parseSystemdEnvAssignment(raw);
                if (parsed) {
                    environment[parsed.key] = parsed.value;
                }
            }
        }
        if (!execStart) {
            return null;
        }
        const programArguments = parseSystemdExecStart(execStart);
        return {
            programArguments,
            ...(workingDirectory ? { workingDirectory } : {}),
            ...(Object.keys(environment).length > 0 ? { environment } : {}),
            sourcePath: unitPath,
        };
    }
    catch {
        return null;
    }
}
export function parseSystemdShow(output) {
    const entries = parseKeyValueOutput(output, "=");
    const info = {};
    const activeState = entries.activestate;
    if (activeState) {
        info.activeState = activeState;
    }
    const subState = entries.substate;
    if (subState) {
        info.subState = subState;
    }
    const mainPidValue = entries.mainpid;
    if (mainPidValue) {
        const pid = Number.parseInt(mainPidValue, 10);
        if (Number.isFinite(pid) && pid > 0) {
            info.mainPid = pid;
        }
    }
    const execMainStatusValue = entries.execmainstatus;
    if (execMainStatusValue) {
        const status = Number.parseInt(execMainStatusValue, 10);
        if (Number.isFinite(status)) {
            info.execMainStatus = status;
        }
    }
    const execMainCode = entries.execmaincode;
    if (execMainCode) {
        info.execMainCode = execMainCode;
    }
    return info;
}
async function execSystemctl(args) {
    try {
        const { stdout, stderr } = await execFileAsync("systemctl", args, {
            encoding: "utf8",
        });
        return {
            stdout: String(stdout ?? ""),
            stderr: String(stderr ?? ""),
            code: 0,
        };
    }
    catch (error) {
        const e = error;
        return {
            stdout: typeof e.stdout === "string" ? e.stdout : "",
            stderr: typeof e.stderr === "string" ? e.stderr : typeof e.message === "string" ? e.message : "",
            code: typeof e.code === "number" ? e.code : 1,
        };
    }
}
export async function isSystemdUserServiceAvailable() {
    const res = await execSystemctl(["--user", "status"]);
    if (res.code === 0) {
        return true;
    }
    const detail = `${res.stderr} ${res.stdout}`.toLowerCase();
    if (!detail) {
        return false;
    }
    if (detail.includes("not found")) {
        return false;
    }
    if (detail.includes("failed to connect")) {
        return false;
    }
    if (detail.includes("not been booted")) {
        return false;
    }
    if (detail.includes("no such file or directory")) {
        return false;
    }
    if (detail.includes("not supported")) {
        return false;
    }
    return false;
}
async function assertSystemdAvailable() {
    const res = await execSystemctl(["--user", "status"]);
    if (res.code === 0) {
        return;
    }
    const detail = res.stderr || res.stdout;
    if (detail.toLowerCase().includes("not found")) {
        throw new Error("systemctl not available; systemd user services are required on Linux.");
    }
    throw new Error(`systemctl --user unavailable: ${detail || "unknown error"}`.trim());
}
export async function installSystemdService({ env, stdout, programArguments, workingDirectory, environment, description, }) {
    await assertSystemdAvailable();
    const unitPath = resolveSystemdUnitPath(env);
    await fs.mkdir(path.dirname(unitPath), { recursive: true });
    const serviceDescription = description ??
        formatGatewayServiceDescription({
            profile: env.OPENCLAW_PROFILE,
            version: environment?.OPENCLAW_SERVICE_VERSION ?? env.OPENCLAW_SERVICE_VERSION,
        });
    const unit = buildSystemdUnit({
        description: serviceDescription,
        programArguments,
        workingDirectory,
        environment,
    });
    await fs.writeFile(unitPath, unit, "utf8");
    const serviceName = resolveGatewaySystemdServiceName(env.OPENCLAW_PROFILE);
    const unitName = `${serviceName}.service`;
    const reload = await execSystemctl(["--user", "daemon-reload"]);
    if (reload.code !== 0) {
        throw new Error(`systemctl daemon-reload failed: ${reload.stderr || reload.stdout}`.trim());
    }
    const enable = await execSystemctl(["--user", "enable", unitName]);
    if (enable.code !== 0) {
        throw new Error(`systemctl enable failed: ${enable.stderr || enable.stdout}`.trim());
    }
    const restart = await execSystemctl(["--user", "restart", unitName]);
    if (restart.code !== 0) {
        throw new Error(`systemctl restart failed: ${restart.stderr || restart.stdout}`.trim());
    }
    // Ensure we don't end up writing to a clack spinner line (wizards show progress without a newline).
    stdout.write("\n");
    stdout.write(`${formatLine("Installed systemd service", unitPath)}\n`);
    return { unitPath };
}
export async function uninstallSystemdService({ env, stdout, }) {
    await assertSystemdAvailable();
    const serviceName = resolveGatewaySystemdServiceName(env.OPENCLAW_PROFILE);
    const unitName = `${serviceName}.service`;
    await execSystemctl(["--user", "disable", "--now", unitName]);
    const unitPath = resolveSystemdUnitPath(env);
    try {
        await fs.unlink(unitPath);
        stdout.write(`${formatLine("Removed systemd service", unitPath)}\n`);
    }
    catch {
        stdout.write(`Systemd service not found at ${unitPath}\n`);
    }
}
export async function stopSystemdService({ stdout, env, }) {
    await assertSystemdAvailable();
    const serviceName = resolveSystemdServiceName(env ?? {});
    const unitName = `${serviceName}.service`;
    const res = await execSystemctl(["--user", "stop", unitName]);
    if (res.code !== 0) {
        throw new Error(`systemctl stop failed: ${res.stderr || res.stdout}`.trim());
    }
    stdout.write(`${formatLine("Stopped systemd service", unitName)}\n`);
}
export async function restartSystemdService({ stdout, env, }) {
    await assertSystemdAvailable();
    const serviceName = resolveSystemdServiceName(env ?? {});
    const unitName = `${serviceName}.service`;
    const res = await execSystemctl(["--user", "restart", unitName]);
    if (res.code !== 0) {
        throw new Error(`systemctl restart failed: ${res.stderr || res.stdout}`.trim());
    }
    stdout.write(`${formatLine("Restarted systemd service", unitName)}\n`);
}
export async function isSystemdServiceEnabled(args) {
    await assertSystemdAvailable();
    const serviceName = resolveSystemdServiceName(args.env ?? {});
    const unitName = `${serviceName}.service`;
    const res = await execSystemctl(["--user", "is-enabled", unitName]);
    return res.code === 0;
}
export async function readSystemdServiceRuntime(env = process.env) {
    try {
        await assertSystemdAvailable();
    }
    catch (err) {
        return {
            status: "unknown",
            detail: String(err),
        };
    }
    const serviceName = resolveSystemdServiceName(env);
    const unitName = `${serviceName}.service`;
    const res = await execSystemctl([
        "--user",
        "show",
        unitName,
        "--no-page",
        "--property",
        "ActiveState,SubState,MainPID,ExecMainStatus,ExecMainCode",
    ]);
    if (res.code !== 0) {
        const detail = (res.stderr || res.stdout).trim();
        const missing = detail.toLowerCase().includes("not found");
        return {
            status: missing ? "stopped" : "unknown",
            detail: detail || undefined,
            missingUnit: missing,
        };
    }
    const parsed = parseSystemdShow(res.stdout || "");
    const activeState = parsed.activeState?.toLowerCase();
    const status = activeState === "active" ? "running" : activeState ? "stopped" : "unknown";
    return {
        status,
        state: parsed.activeState,
        subState: parsed.subState,
        pid: parsed.mainPid,
        lastExitStatus: parsed.execMainStatus,
        lastExitReason: parsed.execMainCode,
    };
}
async function isSystemctlAvailable() {
    const res = await execSystemctl(["--user", "status"]);
    if (res.code === 0) {
        return true;
    }
    const detail = (res.stderr || res.stdout).toLowerCase();
    return !detail.includes("not found");
}
export async function findLegacySystemdUnits(env) {
    const results = [];
    const systemctlAvailable = await isSystemctlAvailable();
    for (const name of LEGACY_GATEWAY_SYSTEMD_SERVICE_NAMES) {
        const unitPath = resolveSystemdUnitPathForName(env, name);
        let exists = false;
        try {
            await fs.access(unitPath);
            exists = true;
        }
        catch {
            // ignore
        }
        let enabled = false;
        if (systemctlAvailable) {
            const res = await execSystemctl(["--user", "is-enabled", `${name}.service`]);
            enabled = res.code === 0;
        }
        if (exists || enabled) {
            results.push({ name, unitPath, enabled, exists });
        }
    }
    return results;
}
export async function uninstallLegacySystemdUnits({ env, stdout, }) {
    const units = await findLegacySystemdUnits(env);
    if (units.length === 0) {
        return units;
    }
    const systemctlAvailable = await isSystemctlAvailable();
    for (const unit of units) {
        if (systemctlAvailable) {
            await execSystemctl(["--user", "disable", "--now", `${unit.name}.service`]);
        }
        else {
            stdout.write(`systemctl unavailable; removed legacy unit file only: ${unit.name}.service\n`);
        }
        try {
            await fs.unlink(unit.unitPath);
            stdout.write(`${formatLine("Removed legacy systemd service", unit.unitPath)}\n`);
        }
        catch {
            stdout.write(`Legacy systemd unit not found at ${unit.unitPath}\n`);
        }
    }
    return units;
}
